pkg/kubelet/pod_workers.go

```diff
@@ -1030,240 +1294,254 @@ func (p *podWorkers) acknowledgeTerminating(pod *v1.Pod) PodStatusFunc {
-			p.removeTerminatedWorker(uid)
+		// we only remove workers that are not represented in our config in order to
+		// prevent accidental restarts of previously terminated pods
+		_, knownPod := known[uid]
+		orphan := !knownPod
+
+		switch {
+		case status.restartRequested:
+			if restartable == nil {
+				restartable = make(map[types.UID]struct{})
+			}
+			restartable[uid] = struct{}{}
+			fallthrough
+		case orphan:
+			if p.removeTerminatedWorker(uid, status, orphan) {
+				continue
+			}
 		}
 		switch {
 		case !status.terminatedAt.IsZero():
 			if status.restartRequested {
-				workers[uid] = TerminatedAndRecreatedPod
-			} else {
-				workers[uid] = TerminatedPod
+				if restartable == nil {
+					restartable = make(map[types.UID]struct{})
+				}
+				restartable[uid] = struct{}{}
 			}
+			workers[uid] = TerminatedPod
 		case !status.terminatingAt.IsZero():
 			workers[uid] = TerminatingPod
 		default:
 			workers[uid] = SyncPod
 		}
 	}
-	return workers
+	return workers, restartable
 }
 
 // removeTerminatedWorker cleans up and removes the worker status for a worker
 // that has reached a terminal state of "finished" - has successfully exited
 // syncTerminatedPod. This "forgets" a pod by UID and allows another pod to be
-// recreated with the same UID.
-func (p *podWorkers) removeTerminatedWorker(uid types.UID) {
-	status, ok := p.podSyncStatuses[uid]
-	if !ok {
-		// already forgotten, or forgotten too early
-		klog.V(4).InfoS("Pod worker has been requested for removal but is not a known pod", "podUID", uid)
-		return
-	}
-
+// recreated with the same UID. The kubelet preserves state about recently
+// terminated pods to prevent accidentally restarting a terminal pod, which is
+// proportional to the number of pods described in the pod config. The method
+// returns true if the worker was completely removed.
+func (p *podWorkers) removeTerminatedWorker(uid types.UID, status *podSyncStatus, orphaned bool) bool {
 	if !status.finished {
-		klog.V(4).InfoS("Pod worker has been requested for removal but is still not fully terminated", "podUID", uid)
-		return
+		// If the pod worker has not reached terminal state and the pod is still known, we wait.
+		if !orphaned {
+			klog.V(4).InfoS("Pod worker has been requested for removal but is still not fully terminated", "podUID", uid)
+			return false
+		}
+
+		// all orphaned pods are considered deleted
+		status.deleted = true
+
+		// When a pod is no longer in the desired set, the pod is considered orphaned and the
+		// the pod worker becomes responsible for driving the pod to completion (there is no
+		// guarantee another component will notify us of updates).
+		switch {
+		case !status.IsStarted() && !status.observedRuntime:
+			// The pod has not been started, which means we can safely clean up the pod - the
+			// pod worker will shutdown as a result of this change without executing a sync.
+			klog.V(4).InfoS("Pod is orphaned and has not been started", "podUID", uid)
```

Comment:
I'm not super clear on this case -- if the pod is not reported as started and we haven't seen it in runtime, is the idea we just wait? Why don't we need to start the termination of it (i.e. similar to the below case)? A little confused on what takes care of the termination in this case, i.e. setting `terminatingAt`?

Reply:
No, we continue past the switch and the worker is removed.  Basically, we didn't start it, and we have no evidence that the pod exists at runtime, so we don't need to invoke any terminating methods because we never called sync.  Also commented above.